DM-17538: Generate stack-produced (rather than CP) calibration products for DECam

mrawls opened the pull request:
This PR enables obs_decam to ingest both CP (community pipeline) calibs and stack-built calibs. Previously we were only supporting CP calibs because nobody had built DECam master calibs with the stack before. I have now successfully constructed biases and (g-band) flats and ingested them.

parejkoj commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -154,18 +167,26 @@ def getDestination(self, butler, info, filename):
+            elif ("illumcor") in calibType.lower():
+                raw = butler.get("cpIllumcor_filename", info)[0]
+            else:
+                assert False, "Invalid CP calibType '{:s}'".format(calibType)
```
`assert False` should be replaced with raising an appropriate exception. `RuntimeError` is a good default, but there may be a better option (ask on slack#dm-butler?). You can also simplify the string to an f-string.

timj commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -135,7 +143,12 @@ def getExtensionName(md):
+            ext = md.getScalar("EXTNAME")
+            return ext
+        except KeyError:
+            return None
```
Use `md.get("EXTNAME", None)` as if it was a normal dict. Then you don't need the try/except.

mrawls replied:
When I don't override it, the code in pipe_tasks' ingest throws a KeyError because `md.getScalar("EXTNAME")` doesn't returns something that is indexable with `[1]` (I pasted the pipe_tasks code below). I can't quite figure out what is happening here.
```
    @staticmethod
    def getExtensionName(md):
        """ Get the name of an extension.
        @param md: PropertySet like one obtained from lsst.afw.fits.readMetadata)
        @return Name of the extension if it exists.  None otherwise.
        """
        try:
            # This returns a tuple
            ext = md.getScalar("EXTNAME")
            return ext[1]
        except lsst.pex.exceptions.Exception:
            return None
```

timj replied:
What happens if you use my suggestion of removing all of that and using `md.get("EXTNAME", None)`?

The related question is why do these raws you are ingesting not have an EXTNAME?

timj replied:
if `ext` is returning a scalar this implies that the EXTNAME string only has one character in it???  When I first looked I assumed it was reading the second element from a list but getScalar doesn't return a list. What is EXTNAME in this file?

mrawls replied:
The EXTNAMEs are things like 'S29       ', 'N4       ', etc. for stack-built MasterCals. It's the CP MasterCals which don't have EXTNAME at all, because they're multi-extension FITS files.

mrawls replied:
`ext = md.get("EXTNAME", None)` seems to work fine, but I'm loathe to put that in pipe_tasks and break non-decam things.

timj replied:
EXTNAME is part of the FITS standard so there's no problem with looking for it. The problem seems to be in requiring the value of EXTNAME to be something useful. I still don't understand how S29 is going to do anything other than return "2" -- is that helpful? That `ext[1]` makes no sense to me at all.

mrawls replied:
Sorry I was mixing up errors here. There are two problems. (1) the pipe_tasks getExtension behavior fails for stack-built calibs because all but the 0th extension lack EXTNAME (it checks every extension). (2) agreed the pipe_tasks getExtension `ext[1]` makes no sense. As best as I can tell the getExtension method was only ever added because of decam. I'll fix this in pipe_tasks.

timj commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -113,9 +114,16 @@ def translate_filter(self, md):
-            if md.exists("OBSTYPE") and "zero" in md.getScalar("OBSTYPE").strip().lower():
-                return "NONE"
-            return CalibsParseTask.translate_filter(self, md)
+            if md.exists("OBSTYPE"):
```
`if "OBSTYPE" in md:`

timj commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -113,9 +114,16 @@ def translate_filter(self, md):
+                    return "NONE"
+            filterName = CalibsParseTask.translate_filter(self, md)
+            # Workaround if/else that can be removed when DM-24514 is complete
+            if filterName == '_unknown_' and md.exists("CALIB_ID"):
```
In most cases you can think of a PropertyList as a dict and use normal dict code. So `"CALIB_ID" in md` and  `md["OBSTYPE"]`

mrawls commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -154,18 +167,26 @@ def getDestination(self, butler, info, filename):
+        md = readMetadata(filename, self.config.hdu)
+        if not md.exists("PROCTYPE"):
+            raise RuntimeError("Unable to find the calib header keyword PROCTYPE in %s, hdu %d" %
+                               (filename, self.config.hdu))
```
Yes; it's assigned by NOAO. Their docs say: "Values include 'Raw', 'InstCal' (a single exposure which has been flux and astrometrically calibrated but not resampled), 'SkySub' (an InstCal frame which has been sky background subtracted), 'Resampled' (a single exposure which has been resampled to a standard orientation), and 'MasterCal' (a master calibration general created from a sequence of exposures)."

mrawls replied:
I've checked a handful of DECam calibs myself and they are all either 'raw' or 'MasterCal'

mrawls commented:
Please take another look @parejkoj. I believe I have addressed your comments here except for the larger issue of test coverage. As best I can tell, there are presently no tests for any ingestion except curated calibs (i.e., defects and friends). I'd be open to pair coding some tests with you, on this ticket or another one.

parejkoj commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -154,18 +153,26 @@ def getDestination(self, butler, info, filename):
+            raise RuntimeError(f"Unable to find the calib header keyword PROCTYPE \
+                               in {filename}, hdu {self.config.hdu}")
+        proctype = md["PROCTYPE"].strip()
+        if "MasterCal" in proctype:  # CP calibration product case
```
Please expand this acronym, since we have our own `cp_pipe`: "DECam Community Pipeline calibration product"

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+
+    Uses a bias+flat pair of LSST-Science-Pipeline-built calibs as well as
+    a bias+flat pair of DECam Community Pipeline calibs, all full of zeros,
+    which live in testdata_decam.
```
Useful docstring, thanks.

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+    which live in testdata_decam.
+    """
+    def setUp(self):
+        self.task = ingestCalibs.DecamCalibsParseTask(name='ItsATest')
```
Is it really? ;-)

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+        # LSST-Science-Pipelines-built bias
+        phuInfo, infoList = self.task.getInfo(self.biasFile)
+        assert phuInfo['calib_hdu'] == 1
+        assert 'path' in phuInfo
```
All of these `assert`s should be turned into `self.assertX` statements (for `X=` e.g. `Equal`, `In`, etc.).

https://developer.lsst.io/python/testing.html

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+        assert 'path' in phuInfo
+        for item in infoList:
+            assert item['calib_hdu'] == 1
+            assert 'path' in item
```
It looks like the above test blocks are the same for both files. Maybe pull those out to `checkGetInfo(self, filename)` which does all the work, and then you just call it twice here, and again below in `testCpCalibsGetInfo`?

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+        flatFilter = self.task.translate_filter(md)
+        assert flatFilter == "g"
+
+    def testCpCalibsGetInfo(self):
```
Ugh, I really don't like this camelCasing, but I guess it's technically correct per our dev guide. grumble_underscores_are_better_grumble...

parejkoj commented on tests/test_parseCalibsGen2.py:
```diff
@@ -0,0 +1,167 @@
+        dataId = {'ccdnum': 1, 'date': '2015-02-18'}
+        filename = self.task.getDestination(self.butler, dataId, self.biasFile)
+        self.assertIn(f"BIAS/{dataId['date']}/BIAS-{dataId['date']}-{dataId['ccdnum']:02}.fits", filename)
+        self.assertNotIn('[', filename)
```
Good one to check for!
